core/ecschema-metadata/src/Validation/SchemaCompareDiagnostics.ts

```diff
@@ -65,6 +65,12 @@ export const SchemaCompareCodes = {
   SchemaReferenceDelta: "SC-128",
 };
 
+// TODO: Remove once 'import() types' are supported by api-extractor. https://github.com/microsoft/rushstack/pull/1916
+import { DiagnosticCategory, DiagnosticType } from "./Diagnostic";
```

Comment:
@calebmshafer I realize there's something tricky going on here, but I just don't get it. Why does dropping this group of unreferenced imports in the middle of the code do anything, and where is the `import()` you removed?

Reply:
@kabentley those types are used as part of the inferred type of Diagnostics so they're add in the `.d.ts` file using the `import()` syntax.  API Extractor uses the `.d.ts` file and comes across them and doesn't effectively handle that unknown concept.

The unreferenced imports force Typescript to resolve those imports like normal in the `d.ts` file which removes any of the `import()` appearances.

API Extractor is working on support for `import()` but it's been stalled for a while.  Hopefully they pick it back up soon and we can remove this.

Reply:
I don't understand what you mean by "inferred type of Diagnostics". Seems like it would be easier to just disable extract-api for this package.

Reply:
The `SchemaCompareDiagnostics` variable has a complicated type due to each property of it being a new class type  with the parameters provided. For example, `SchemaCompareDiagnostics.SchemaReferenceMissing` has the following type,

```ts
SchemaReferenceMissing: {
        new (schema: Schema, messageArgs: [Schema], category?: DiagnosticCategory): {
            readonly code: string;
            readonly messageText: string;
            readonly schema: Schema;
            readonly diagnosticType: DiagnosticType;
            ecDefinition: Schema;
            messageArgs?: [Schema] | undefined;
            category: DiagnosticCategory;
        };
        code: string;
        diagnosticType: DiagnosticType;
    };
```

We use this in order to gather a list of Diagnostics, as a different type for each, that can be output as part of the validation of the Schema, it's a similar approach to how the Typescript parser works when building up your list of errors.

We could switch to turn it off but I was trying to avoid it for the whole package.  `import()` is actually used quite often in the whole mono-repo and api-extractor by-passes it right now without warning or failure.  In future versions of API Extractor they actually make it an error which is preventing us from moving up versions until they add full support.

Reply:
frankly, i don't understand why anything about this package is public. Who would use it?

Reply:
We use some of the classes in the `imodeljs-backend` package for the @scsewall's Transformer.

The Diagnostic classes are specifically used for BIS schema validation and the Schema Editor at this point.  When the Schema Editor starts actually editing that's when it will be heavily used in the app.

Reply:
I can see why *we* use it. In both cases above consumers should never see it. This very much seems like an @internal package to me.

Reply:
To clarify, the usage in `imodeljs-backend` is part of the `IModelExportHandler` class on the `shouldExportSchema` and `onExportSchema` methods.  Both of which are intended to be overridden to implement your specific logic when writing the transformation so any transformer (and potentially Connectors in the future) will use most of the classes in the package.

There are two other groups of functionality outside of the base data model, deserializing/serializing schemas from both JSON and XML and the Validation code against both EC and BIS rules.

The JSON deserialization functionality is the only place JSON is currently supported (not supported in the native code), which I think is used by Schema Service and subsequently Plant and Insights.

The Validation code is against both EC and BIS rules and most likely only be internally used unless a 3rd party has a reason to use it.  However, once editing starts happening more often being able to customize your Schema via a UI, or even manually, may make the validation APIs more relevant to give good errors messages in various apps (mostly thinking Plant in this case).

Reply:
Our sample Connector uses it for dynamic schema generation.  https://github.com/imodeljs/itwin-connector-sample/blob/main/COBie-connector/src/DynamicSchemaGenerator.ts